syntax = "proto3";

package structuredb.v1;

message UpsertTableRequest {
  optional string tx = 1;
  string table = 2;
  string key = 3;
  string value = 4;
}

message UpsertTableResponse {
  string tx = 1;
}

message LookupTableRequest {
  optional string tx = 1;
  string table = 2;
  string key = 3;
}

message LookupTableResponse {
  optional string value = 1;
}

message CreateTableRequest {
  optional string tx = 1;
  string name = 2;
}

message CreateTableResponse {
  string tx = 1;
}

message DropTableRequest {
  optional string tx = 1;
  string name = 2;
}

message DropTableResponse {
  string tx = 1;
}

message DeleteTableRequest {
  optional string tx = 1;
  string table = 2;
  string key = 3;
}

message DeleteTableResponse {
  string tx = 1;
}

message GetAllTableRequest {
  optional string tx = 1;
  string table = 2;
}

message Record {
  string key = 1;
  string value = 2;
};

message GetAllTableResponse {
  string tx = 1;
  repeated Record records = 2;
}

service Tables {
  rpc Upsert(UpsertTableRequest) returns (UpsertTableResponse) {}
  rpc Lookup(LookupTableRequest) returns (LookupTableResponse) {}
  rpc GetAll(GetAllTableRequest) returns (GetAllTableResponse) {}
  rpc Delete(DeleteTableRequest) returns (DeleteTableResponse) {}
  rpc CreateTable(CreateTableRequest) returns (CreateTableResponse) {}
  rpc DropTable(DropTableRequest) returns (DropTableResponse) {}
}
